### ISSUE TITLE
[Proposal] Dynamic Localization

### ISSUE THREAD

**Issue by tsteinholz** [SFML/SFML-Website#47]
Before I get on my spiel about this proposal, I would like to mention it will be a pain in the ass to implement, but in the end, worth it.

So, the way you guys have language set up, you basically have either French or English, and each page has to be in the folder twice. Once in English, once in French. Every change to each page you have to do twice. This is tedious redundant work that could be avoided with a proper dynamic system.

My solution to this problem is not ultimate, it's not the only answer, and its not necessarily the right answer, but hear me out. So when I had to do something like this for my website, I basically set up a dictionary service in php. This service is actually really simple, the steps for implementation include
 * Create "page_name"-"language-abbreviation"."json / xml / php / etc"
 * Make each webpage <page_name>.php load in the correct language
For this step, I found the easiest way being java-script cookies, defaulting to english.

Check out what I did for [my website](https://github.com/LastStandStudio/Website) as a reference to what I am talking about, and this is not a perfect system, it could defiantly use improvements in how everything is set up. And you could tweak those things when you are implementing them. 

So basically the end result will be one .php webpage that loads in all the strings from some sort of data file based on the cookie setting for language (or whatever other way you want to detect/set language). Then for each webpage you can have an infinite amount of language options based on what data files you have. Now at first you could have English and French transferred over to the new data file, then depend on the community to translate the rest to what ever language they would like to see the website support.

**Comment by Bromeon** [comment]
And how would that prevent us from having to make changes twice (to English and French)?

**Comment by LaurentGomila** [comment]
His idea is basically to split what's common (the page structure) and what's specific to each language (pure text), instead of treating (duplicating) everything as a block.

This is of course a good idea, and the basis for multilingual systems, but I wonder if this would really be relevant for SFML website, where most of the pages content is text (I especially think about tutorials). Decoupling text and structure could even make things worse, because it would become more difficult to follow the page flow to make modifications to tutorials.

**Comment by tsteinholz** [comment]
With the documentation, since the pages are mostly text it might seem at first a little redundant, but it makes the code much cleaner and easier to modify. Like you said earlier, the little formatting that you would have to do in a text only file, you would only have to do once. So to make modifications you change the php, then do what you need to each language file. This is a more robust system then making the changes, copying everything to a new document, then beginning to translate. Code wise it's redundant because you have to write the same formatting twice which can lead to human error. This system might be a bit overkill for what you guys are using it for, but in the end if you support more languages, more people will be attracted to using your software.

**Comment by eXpl0it3r** [comment]
It's common practice and it allows for more flexibility, we're well aware of that, however there is little to none gain for us. While "supporting more languages" sounds like a nice feature on paper, it's a lot more work to maintain it.

We are now already facing minor delays when adjusting or adding something, because someone has to translate the text to French as well. Fortunately we have two active team members who are fluent in French, so the delays really are just minor, but think about what happens if I want to add one paragraph and then have to wait for 10 people to translate it into their own language.

You have to keep in mind that SFML is not an end-user software, it's a library for software developers and fact is that if you are a software developer and can't understand English, you'll be facing issues everywhere and not just reading the SFML tutorials. Additionally even if we did translate to more languages, many things will always be available in English only (e.g. documentation, guidelines, change logs, API names, etc.). And finally providing more languages will imply for some users that we'll be giving support in said language, which of course is impossible when we don't even speak the language and it may bring frustration onto us and these users.

**Comment by tsteinholz** [comment]
I agree with you completely, all I am saying is that since you guys already support French on your home page, you should do it right. Thus leaving the door open for other languages (like German as you mentioned). You're right in that writing all the documentation, guidelines, etc. in every language is extremely excessive and was not what I was proposing. I'm defiantly not suggesting you need to support any other language more than you do french already. The only text that should be loaded dynamically is the text you already have and use for the french pages. Yes this isn't a high priority especially if you guys are having minor delays, but it is worth having in mind for the future.

**Comment by eXpl0it3r** [comment]
Okay :blush: 

As I mentioned right now there's little to non gain in changing anything. It's work as it is right now and having to convert everything will only generate a lot of work. But if we ever decide to add a new language, I think we'll certainly think about it again.

For now I'll just close this issue, since it's not something we're actively planing for.

**Comment by mantognini** [comment]
I agree, with the current situation the pragmatic reward for implementing such feature is actually negative. Well reconsider our position the day we have more languages to support. ;)

**Comment by tsteinholz** [comment]
Would you guys care for me to fork the website and create my own solution for my proposal so that you guys don't have to be burdened with the task while you guys are busy? Or do you not care at all to have it implemented at this time.

**Comment by Bromeon** [comment]
Implementation effort was not a counterargument in the discussion above, but *maintenance* effort was. That is, even if somebody else implements it, it would cause more work.

**Comment by tsteinholz** [comment]
The point was to decrease maintenance effort with a working system that will take implementation effort that I was willing to contribute, but I won't implement it if you don't care for the system.

**Comment by mantognini** [comment]
It's not that we don't care, not at all. We highly appreciate the contribution offer. ;)

But the fact is that having such system: a) increases the completely of the code (because of indirections, more code and technologies are involved) and b) we need to learn it to update the site's content.

If you want to contribute on something related to the website, there was a discussion a few weeks ago about adding a showcase page. No decision was actually made as far as I can remember but you could revive the conversation and make a proposal. :) http://en.sfml-dev.org/forums/index.php?topic=18553.0

**Comment by tsteinholz** [comment]
Thanks for the feedback @mantognini , I am more of a low-level, back-end guy. I don't think I would be the best person for the job. If it is critical and nobody else has time, I could do it. But then again I don't think I'm the best person for the job.
